packages/compat/src/transaction.ts

```diff
@@ -0,0 +1,141 @@
+    VersionedTransaction,
+} from '@solana/web3.js';
+
+import { SignedVersionedTransaction } from './signed-versioned-transaction';
+
+function convertAccount(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    accountIndex: number
+): IAccountMeta {
+    const accountPublicKey = accountKeys.get(accountIndex);
+    if (!accountPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find account address at index ${accountIndex}`);
+    }
+    const isSigner = message.isAccountSigner(accountIndex);
+    const isWritable = message.isAccountWritable(accountIndex);
+
+    const role = isSigner
+        ? isWritable
+            ? AccountRole.WRITABLE_SIGNER
+            : AccountRole.READONLY_SIGNER
+        : isWritable
+        ? AccountRole.WRITABLE
+        : AccountRole.READONLY;
+
+    return {
+        address: accountPublicKey.toBase58() as Base58EncodedAddress,
+        role,
+    };
+}
+
+function convertInstruction(
+    message: VersionedMessage,
+    accountKeys: MessageAccountKeys,
+    instruction: MessageCompiledInstruction
+): IInstruction {
+    const programAddressPublicKey = accountKeys.get(instruction.programIdIndex);
+    if (!programAddressPublicKey) {
+        // TODO coded error
+        throw new Error(`Could not find program address at index ${instruction.programIdIndex}`);
+    }
+
+    const accounts = instruction.accountKeyIndexes.map(accountIndex =>
+        convertAccount(message, accountKeys, accountIndex)
+    );
+
+    return {
+        programAddress: programAddressPublicKey.toBase58() as Base58EncodedAddress,
+        ...(accounts.length ? { accounts } : {}),
+        ...(instruction.data.length ? { data: instruction.data } : {}),
+    };
+}
+
+function convertSignatures(
+    transaction: VersionedTransaction,
+    staticAccountKeys: PublicKey[]
+): ITransactionWithSignatures['signatures'] {
+    return transaction.signatures.reduce((acc, sig, index) => {
+        // old web3js includes a fake all 0 signature if it hasn't been signed
+        // we don't do that for the new tx model. So just skip if it's all 0s
+        const allZeros = sig.every(byte => byte === 0);
+        if (allZeros) return acc;
+
+        const address = staticAccountKeys[index].toBase58() as Base58EncodedAddress;
+        return { ...acc, [address]: sig as Ed25519Signature };
+    }, {});
+}
+
+export function fromOldVersionedTransactionWithBlockhash(
+    transaction: SignedVersionedTransaction,
+    lastValidBlockHeight: bigint
+): Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime & ITransactionWithSignatures;
+
+export function fromOldVersionedTransactionWithBlockhash(
+    transaction: VersionedTransaction,
+    lastValidBlockHeight: bigint
+): Transaction & ITransactionWithFeePayer & ITransactionWithBlockhashLifetime;
+
+export function fromOldVersionedTransactionWithBlockhash(
+    transaction: VersionedTransaction,
+    lastValidBlockHeight: bigint
```

Comment:
Ahhh this is annoying lol. I almost wonder what would be better: Requiring you to hang onto `lastValidBlockHeight` whenever you added the blockhash to your old tx, or simply passing both here - `blockhash` and `lastValidBlockHeight`

Reply:
Yep fair point! Might be better to just have you pass both. I guess otherwise you have to mutate the old transaction to have the matching blockhash.